### PR TITLE
Remove all dependencies on exp-verbs

### PR DESCRIPTION
**Pull request by pakmarkthub**
**This change:**
- addresses issue #79.
- removes exp-verbs and replaces with directverbs.
- enables libgdsync to be compatible with ib-verbs (non-exp) from the perspective of users.
- fixes issue #83.
- aligns code in all files.
- fixes issue #88 and #89.

**Presubmit Testings:**
- gc11 and PSG Cluster
- Tests: gds_kernel_latency, gds_kernel_loopback_latency

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -144,12 +144,12 @@ dnl Checks for Verbs support
 AC_CHECK_LIB(ibverbs, ibv_get_device_list, [],
     AC_MSG_ERROR([ibv_get_device_list() not found.  libgdsync requires libibverbs.]))
 
-AC_CHECK_LIB(ibverbs, ibv_exp_create_qp,
-    AC_MSG_ERROR([ibv_exp_create_qp not found. libgdsync requires verbs extension support.]))
-
-AC_CHECK_HEADER(infiniband/peer_ops.h, [],
-    AC_MSG_ERROR([<infiniband/peer_ops.h> not found.  libgdsync requires verbs peer-direct support.]))
-AC_HEADER_STDC
+#AC_CHECK_LIB(ibverbs, ibv_exp_create_qp,
```
Comment (drossetti) [inline configure.ac]: you should get rid of these altogether

File: include/gdsync/core.h [hunk 2]
```diff
@@ -39,37 +39,87 @@
...
+
+        unsigned int sq_cur_post;
+        unsigned int bf_offset;
+
+        uint32_t peer_scur_post;
+        uint64_t peer_va_id_dbr;
+        uint64_t peer_va_id_bf;
 };
 
+int gds_poll_cq(struct gds_cq *cq, int ne, struct ibv_wc *wc);
```
Comment (drossetti) [inline include/gdsync/core.h]: why do you need this here? isn't it a public API?

File: include/gdsync/core.h [hunk 3]
```diff
@@ -39,37 +39,87 @@
...
+} gds_cq_type_t;
 
 struct gds_cq {
-        struct ibv_cq *cq;
-        uint32_t curr_offset;
+        struct ibv_cq          *cq;
+        uint32_t                curr_offset;
+        uint32_t                cons_index;
+        gds_cq_type_t           type;
+
```
Comment (drossetti) [inline include/gdsync/core.h]: please get rid of extra empty lines

File: include/gdsync/core.h [hunk 4]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
-        GDS_MEMORY_IO   = 4,
-	GDS_MEMORY_MASK = 0x7
-} gds_memory_type_t;
-
 // Note: those flags below must not overlap with gds_memory_type_t
 typedef enum gds_wait_flags {
-        GDS_WAIT_POST_FLUSH_REMOTE = 1<<3, /*< add a trailing flush of the ingress GPUDirect RDMA data path on the GPU owning the stream */
-        GDS_WAIT_POST_FLUSH = GDS_WAIT_POST_FLUSH_REMOTE /*< alias for backward compatibility */
+        /** 
+         * add a trailing flush of the ingress GPUDirect RDMA data path on the
```
Comment (drossetti) [inline include/gdsync/core.h]: by reformatting this comment, you are breaking the doxygen doc format.

File: include/gdsync/core.h [hunk 5]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
+        /* add a heading memory barrier to the write value operation */
+        GDS_WRITE_PRE_BARRIER_SYS = 1<<4, 
+        /* alias for backward compatibility */
+        GDS_WRITE_PRE_BARRIER = GDS_WRITE_PRE_BARRIER_SYS 
 } gds_write_flags_t;
 
 typedef enum gds_write_memory_flags {
-        GDS_WRITE_MEMORY_POST_BARRIER_SYS = 1<<4, /*< add a trailing memory barrier to the memory write operation */
-        GDS_WRITE_MEMORY_PRE_BARRIER_SYS  = 1<<5 /*< add a heading memory barrier to the memory write operation, for convenience only as not a native capability */
+        /* add a trailing memory barrier to the memory write operation */
```
Comment (drossetti) [inline include/gdsync/core.h]: same here

File: include/gdsync/core.h [hunk 6]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
+        GDS_WRITE_MEMORY_PRE_BARRIER_SYS  = 1<<5 
 } gds_write_memory_flags_t;
 
 typedef enum gds_membar_flags {
         GDS_MEMBAR_FLUSH_REMOTE = 1<<4,
         GDS_MEMBAR_DEFAULT      = 1<<5,
         GDS_MEMBAR_SYS          = 1<<6,
-        GDS_MEMBAR_MLX5         = 1<<7 /*< modify the scope of the barrier, for internal use only */
+        /* modify the scope of the barrier, for internal use only */
+        GDS_MEMBAR_MLX5         = 1<<7
```
Comment (drossetti) [inline include/gdsync/core.h]: same here

File: include/gdsync/core.h [hunk 7]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
+        GDS_MEMBAR_MLX5         = 1<<7 
 } gds_membar_flags_t;
 
 enum {
         GDS_SEND_INFO_MAX_OPS = 32,
         GDS_WAIT_INFO_MAX_OPS = 32
 };
 
+typedef enum gds_memory_type {
+        /* use this flag for both cudaMalloc/cuMemAlloc and cudaMallocHost/cuMemHostAlloc */
```
Comment (drossetti) [inline include/gdsync/core.h]: moving this enum declaration here adds useless noise

**Comment by drossetti** [comment]
@haggaie this is the PR where we get rid of exp verbs. 
In practice a lot of the infrastructure is pulled from the peer-direct async implementation. 
Can you ack that this ok for you?

File: include/gdsync/core.h [hunk 8]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
+        GDS_MEMORY_GPU  = 1, 
+        GDS_MEMORY_HOST = 2,
+        GDS_MEMORY_IO   = 4,
+        GDS_MEMORY_MASK = 0x7
+} gds_memory_type_t;
+
+enum ibv_exp_peer_direct_attr_mask {
```
Comment (drossetti) [inline include/gdsync/core.h]: this must be an oversight

File: include/gdsync/core.h [hunk 9]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
+        GDS_PEER_DIRECT_VERSION = (1 << 0) /* Must be set */
+};
+
+enum gds_peer_op {
```
Comment (drossetti) [inline include/gdsync/core.h]: the stuff taken from peer_op.h should be moved to an internal header file as it is not part of the public APIs.

File: include/gdsync/core.h [hunk 10]
```diff
@@ -113,62 +163,159 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
...
 /**
  * Represents a posted send operation on a particular QP
  */
 
 typedef struct gds_send_request {
-        struct ibv_exp_peer_commit commit;
-        struct peer_op_wr wr[GDS_SEND_INFO_MAX_OPS];
+        struct gds_peer_commit commit;
+        struct gds_peer_op_wr wr[GDS_SEND_INFO_MAX_OPS];
 } gds_send_request_t;
```
Comment (drossetti) [inline include/gdsync/core.h:185]: we should consider turning gds_send_request_t into an opaque object

File: src/apis.cpp [hunk 11]
```diff
@@ -83,48 +75,13 @@ static void gds_init_wait_request(gds_wait_request_t *request, uint32_t offset)
...
         request->peek.entries = sizeof(request->wr)/sizeof(request->wr[0]);
-        request->peek.whence = IBV_EXP_PEER_PEEK_ABSOLUTE;
+        request->peek.whence = GDS_PEER_PEEK_ABSOLUTE;
         request->peek.offset = offset;
         gds_init_ops(request->peek.storage, request->peek.entries);
 }
 
 //-----------------------------------------------------------------------------
 
-static int gds_rollback_qp(struct gds_qp *qp, gds_send_request_t * send_info, enum ibv_exp_rollback_flags flag)
```
Comment (drossetti) [inline src/apis.cpp:93]: why is rollback_qp not needed anymore?

File: src/apis.cpp [hunk 12]
```diff
@@ -171,32 +123,133 @@ int gds_post_recv(struct gds_qp *qp, struct ibv_recv_wr *wr, struct ibv_recv_wr
...
         assert(qp->qp);
-        ret = ibv_exp_post_send(qp->qp, p_ewr, bad_ewr);
-        if (ret) {
 
-                if (ret == ENOMEM) {
-                        // out of space error can happen too often to report
-                        gds_dbg("ENOMEM error %d in ibv_exp_post_send\n", ret);
-                } else {
-                        gds_err("error %d in ibv_exp_post_send\n", ret);
+        // Emulating ibv_exp_post_send
```
Comment (drossetti) [inline src/apis.cpp]: I believe it'd be cleaner to move the mlx5 implementation of ibv_exp_post_send, say called mlx5_post_send(), into a new cpp module, where you collect all the mlx5 stuff.

File: src/apis.cpp [hunk 13]
```diff
@@ -272,82 +325,87 @@ int gds_stream_post_send_all(CUstream stream, int count, gds_send_request_t *req
...
-                goto out;
-        } else if (retcode) {
-                gds_err("error %d in peer_peek_cq\n", retcode);
-                goto out;
-        }
-        //gds_dump_wait_request(request, 1);
-        out:
-	       return retcode;
-}
+        // Emulating ibv_exp_peer_peek_cq
```
Comment (drossetti) [inline src/apis.cpp]: same here, let's refactor this code and introduce mlx5_peer_peek_cq()

File: src/apis.cpp [hunk 14]
```diff
@@ -387,16 +443,12 @@ int gds_stream_wait_cq(CUstream stream, struct gds_cq *cq, int flags)
         ret = gds_stream_post_wait_cq(stream, &request);
         if (ret) {
                 gds_err("error %d in gds_stream_post_wait_cq_ex\n", ret);
-                int retcode2 = gds_abort_wait_cq(cq, &request);
```
Comment (drossetti) [inline src/apis.cpp:390]: why is aborting not necessary anymore?

File: src/apis.cpp [hunk 15]
```diff
@@ -692,131 +744,119 @@ int gds_post_descriptors(size_t n_descs, gds_descriptor_t *descs, int flags)
         for(i = 0; i < n_descs; ++i) {
                 gds_descriptor_t *desc = descs + i;
                 switch(desc->tag) {
-                case GDS_TAG_SEND: {
-                        gds_dbg("desc[%zu] SEND\n", i);
-                        gds_send_request_t *sreq = desc->send;
-                        retcode = gds_post_ops_on_cpu(sreq->commit.entries, sreq->commit.storage, flags);
```
Comment (drossetti) [inline src/apis.cpp:698]: why are you getting rid of the CPU synchronous path implementation?

File: src/archutils.h [hunk 16]
```diff
@@ -1,4 +1,29 @@
-// TODO: ADD LICENCE HERE
+/* Copyright (c) 2016 - 2020, NVIDIA CORPORATION. All rights reserved.
```
Comment (drossetti) [inline src/archutils.h:1]: I don't believe we should add a copyright and license header here.

File: src/gdsync.cpp [hunk 17]
```diff
@@ -1678,142 +1611,264 @@ gds_peer *peer_from_stream(CUstream stream)
...
-        gcq->cq = ibv_exp_create_cq(context, cqe, cq_context, channel, comp_vector, &attr);
+        gcq->cq = ibv_create_cq(context, cqe, cq_context, channel, comp_vector);
         if (!gcq->cq) {
-            gds_err("error %d in ibv_exp_create_cq, old errno %d\n", errno, old_errno);
-            return NULL;
+                gds_err("error %d in mlx5dv_create_cq, old errno %d\n", errno, old_errno);
+                return NULL;
+        }
+
+        old_errno = errno;
```
Comment (drossetti) [inline src/gdsync.cpp]: can we refactor this code into a call to a mlx5 specific function?

File: src/gdsync.cpp [hunk 18]
```diff
@@ -1678,142 +1611,264 @@ gds_peer *peer_from_stream(CUstream stream)
...
+                .length         = sizeof(struct gds_peek_entry) * gcq->dv_cq.cqe_cnt,
+                .dir            = gcq->peer_dir,
+                .peer_id        = peer_attr->peer_id,
+                .alignment      = (uint32_t)sysconf(_SC_PAGESIZE),
+                .comp_mask      = 0
+        };
+        peer_buf = peer_attr->buf_alloc(&ba_attr);
+        if (!peer_buf) {
+                gds_err("error %d in buf_alloc, old_errno %d\n", errno, old_errno);
+                return NULL;
```
Comment (drossetti) [inline src/gdsync.cpp]: here and in other error cases we are leaking memory or keeping resources registered

File: src/gdsync.cpp [hunk 19]
```diff
@@ -1678,142 +1611,264 @@ gds_peer *peer_from_stream(CUstream stream)
...
-        else
-            gds_warn("NOT using res_domain\n");
+int gds_poll_cq(struct gds_cq *cq, int ne, struct ibv_wc *wc)
+{
+        int p_ne;
+        unsigned int idx;
+        int i = 0;
+        int cnt;
+
+        p_ne = ibv_poll_cq(cq->cq, ne, wc);
```
Comment (drossetti) [inline src/gdsync.cpp]: as a baseline, for gds cq, ibv_poll_cq should only be called after verifying that the peek entry is not busy.
let's forget about the case of errors for the moment.

File: src/gdsync.cpp [hunk 20]
```diff
@@ -1678,142 +1611,264 @@ gds_peer *peer_from_stream(CUstream stream)
...
+                default:
+                        gds_err("request allocation with unsupported resource_type\n");
+                        return NULL;
         }
 
-        return gcq;
+        if (!buf) {
+                int err;
+                gds_dbg("alloc on host\n");
+                if ((err = posix_memalign(&ptr, alignment, size)) != 0) {
```
Comment (drossetti) [inline src/gdsync.cpp]: For the records, by doing so, we prevent libmlx5 from doing optimizations like using contig memory.
To let libmlx5 do its things, we could return IBV_ALLOCATOR_USE_DEFAULT, but then we should make a call to register_va() later.
can you add a comment here?

File: src/gdsync.cpp [hunk 21]
```diff
@@ -1878,21 +1937,54 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context,
...
         gqp->send_cq.cq = qp->send_cq;
         gqp->send_cq.curr_offset = 0;
+        gqp->send_cq.type = GDS_CQ_TYPE_SQ;
+
+        memcpy(&gqp->recv_cq, rx_gcq, sizeof(struct gds_cq));
         gqp->recv_cq.cq = qp->recv_cq;
         gqp->recv_cq.curr_offset = 0;
+        gqp->recv_cq.type = GDS_CQ_TYPE_RQ;
+
+        dv_obj = {
```
Comment (drossetti) [inline src/gdsync.cpp]: can we refactor the mlx5 specific part into a call into a new NIC specific function?

File: src/gdsync_debug_hostregister_bug.cpp [hunk 22]
```diff
@@ -964,91 +964,91 @@ int gds_post_pokes(CUstream stream, int count, gds_send_request_t *info, uint32_
...
+                break;
+            }
+            case IBV_PEER_OP_STORE_QWORD: {
+                uint64_t *ptr = (uint64_t*)((ptrdiff_t)range_from_id(op->wr.qword_va.target_id)->va + op->wr.qword_va.offset);
+                uint64_t data = op->wr.qword_va.data;
+                ACCESS_ONCE(*ptr) = data;
+                gds_dbg("%p <- %016"PRIx64"\n", ptr, data);
+                break;
+            }
+            case IBV_PEE_OP_COPY_BLOCK: {
```
Comment (drossetti) [inline src/gdsync_debug_hostregister_bug.cpp]: _PEE ? :)

File: src/mem.cpp [hunk 23]
```diff
@@ -163,20 +162,31 @@ static int gds_unmap_gdr_memory(gds_mem_desc_t *desc)
 static int gds_alloc_gdr_memory(gds_mem_desc_t *desc, size_t size, int flags)
 {
         CUdeviceptr d_buf = 0;
-        size_t buf_size = size;
+        CUdeviceptr d_buf_aligned = 0;
+        off_t offset = 0;
+        size_t buf_size = size + GDS_GPU_PAGE_SIZE;
```
Comment (drossetti) [inline src/mem.cpp]: -1 missing ?

File: src/mem.cpp [hunk 24]
```diff
@@ -163,20 +162,31 @@ static int gds_unmap_gdr_memory(gds_mem_desc_t *desc)
...
         CUCHECK(cuMemAlloc(&d_buf, buf_size));
-        gds_dbg("allocated GPU polling buffer d_buf=%p\n", (void*)d_buf);
-        //CUCHECK(cuMemsetD8(d_buf, 0, buf_size));
 
-        ret = gds_map_gdr_memory(desc, d_buf, buf_size, flags);
+        offset = d_buf & GDS_GPU_PAGE_OFF;
+        if (offset == 0)
+                d_buf_aligned = d_buf;
+        else
+                d_buf_aligned = (d_buf + GDS_GPU_PAGE_SIZE - 1) & GDS_GPU_PAGE_MASK;
```
Comment (drossetti) [inline src/mem.cpp]: it is safe to get rid of offset calculation and to always do the rounding

File: src/gdsync.cpp [hunk 25]
```diff
@@ -1678,142 +1611,264 @@ gds_peer *peer_from_stream(CUstream stream)
...
+                return p_ne;
+
+        for (cnt = 0; cnt < p_ne; ++cnt) {
+                idx = cq->cons_index & (cq->dv_cq.cqe_cnt - 1);
+                while (cq->peer_peek_table[idx]) {
+                        struct gds_peek_entry *tmp;
+                        while (cq->peer_peek_table[idx]->busy) {
+                                ++i;
+                                if (i >= 16) {
+                                        asm volatile("pause\n": : :"memory");
```
Comment (drossetti) [inline src/gdsync.cpp]: P9 should break, shouldn't you call an appropriate wrapper here?

**Review by drossetti — CHANGES_REQUESTED** [review]
partially reviewed

File: include/gdsync/core.h (hunk 2 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: This is a public API. That's why I put it in that header file, similar to gds_create_qp.

File: include/gdsync/core.h (hunk 7 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: This enum declaration is originally from this file (see line 116 of the original commit). I am moving it back to the same position.

File: include/gdsync/core.h (hunk 10 above)
Comment (pakmarkthub) [reply include/gdsync/core.h:185]: gds_send_request_t is currently allocated by the user (see work_desc_t in gds_kernel_loopback_latency). I suggest that we do it in another PR.

File: include/gdsync/core.h (hunk 5 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Reverted

File: include/gdsync/core.h (hunk 4 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Reverted

File: include/gdsync/core.h (hunk 3 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Done

File: configure.ac (hunk 1 above)
Comment (pakmarkthub) [reply configure.ac]: Removed

File: include/gdsync/core.h (hunk 6 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Reverted

File: include/gdsync/core.h (hunk 8 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Removed

File: include/gdsync/core.h (hunk 9 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: The following are used in gds_send_request_t. 
1. struct gds_peer_commit
2. struct gds_peer_op_wr
3. enum gds_peer_op

We need to turn gds_send_request_t to an opaque object and modify applications (e.g., gds_kernel_loopback_latency) first before we can moved these data structures to an internal header file.

File: include/gdsync/core.h [hunk 26]
```diff
@@ -39,37 +39,86 @@
...
+        GDS_CREATE_QP_WQ_ON_GPU    = 1<<0,
+        GDS_CREATE_QP_TX_CQ_ON_GPU = 1<<1,
+        GDS_CREATE_QP_RX_CQ_ON_GPU = 1<<2,
+        GDS_CREATE_QP_WQ_DBREC_ON_GPU = 1<<5,
+};
+
+typedef struct ibv_qp_init_attr gds_qp_init_attr_t;
+typedef struct ibv_send_wr gds_send_wr;
+
+struct gds_peek_entry {
```
Comment (drossetti) [inline include/gdsync/core.h]: this def does not belong to the public API

Comment (drossetti) [reply include/gdsync/core.h]: you just need a fwd decl

File: include/gdsync/core.h [hunk 27]
```diff
@@ -39,37 +39,86 @@
...
+typedef enum gds_cq_type {
+        GDS_CQ_TYPE_SQ,
+        GDS_CQ_TYPE_RQ
+} gds_cq_type_t;
 
 struct gds_cq {
-        struct ibv_cq *cq;
-        uint32_t curr_offset;
+        struct ibv_cq          *cq;
+        uint32_t                curr_offset;
```
Comment (drossetti) [inline include/gdsync/core.h:74]: do we really need to expose the internal implementation? 
for example, by using struct mlx5dv_cq, you introduce a compile time dependency on the user code

File: include/gdsync/core.h [hunk 28]
```diff
@@ -39,37 +39,86 @@
...
+        uint32_t                peer_dir;
+        struct {
+                void           *buf;
+                size_t          length;
+        } peer_buf;
+        struct gds_peek_entry **peer_peek_table;
+        struct gds_peek_entry  *peer_peek_free;
 };
 
+#define GDS_LAST_PEEK_ENTRY (-1U)
```
Comment (drossetti) [inline include/gdsync/core.h]: this belongs to the implementation side

File: include/gdsync/core.h [hunk 29]
```diff
@@ -39,37 +39,86 @@
...
+        (pe == NULL ? GDS_LAST_PEEK_ENTRY : \
+         ((pe - (struct gds_peek_entry *)cq->peer_buf.buf)))
+
 struct gds_qp {
         struct ibv_qp *qp;
         struct gds_cq send_cq;
         struct gds_cq recv_cq;
-        struct ibv_exp_res_domain * res_domain;
         struct ibv_context *dev_context;
+        struct mlx5dv_qp dv_qp;
```
Comment (drossetti) [inline include/gdsync/core.h]: these are again implementation details

File: include/gdsync/core.h (hunk 2 above)
Comment (drossetti) [reply include/gdsync/core.h]: sorry, I mistakenly thought I was reviewing in a different file

File: include/gdsync/core.h [hunk 30]
```diff
@@ -114,10 +163,11 @@ int gds_stream_queue_send(CUstream stream, struct gds_qp *qp, gds_send_wr *p_ewr
 // batched submission APIs
 
 typedef enum gds_memory_type {
-        GDS_MEMORY_GPU  = 1, /*< use this flag for both cudaMalloc/cuMemAlloc and cudaMallocHost/cuMemHostAlloc */
+        /* use this flag for both cudaMalloc/cuMemAlloc and cudaMallocHost/cuMemHostAlloc */
```
Comment (drossetti) [inline include/gdsync/core.h]: please move this comment back inline for doxygen, one day we will enable it

File: include/gdsync/core.h (hunk 9 above)
Comment (drossetti) [reply include/gdsync/core.h]: ok  that can be done in a follow up change

File: include/gdsync/core.h (hunk 10 above)
Comment (drossetti) [reply include/gdsync/core.h:185]: agreed

Comment (drossetti) [reply include/gdsync/core.h:185]: still, can we move all mlx5/peer-direct specific defs in a separate file?

File: src/apis.cpp (hunk 11 above)
Comment (drossetti) [reply src/apis.cpp:93]: did you address this concern?

**Review by drossetti — CHANGES_REQUESTED** [review]

File: src/gdsync.cpp (hunk 19 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: Fixed

File: src/gdsync.cpp (hunk 20 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: Added a TODO comment in the code. Let me change the mechanism later.

File: src/mem.cpp (hunk 23 above)
Comment (pakmarkthub) [reply src/mem.cpp]: Fixed

File: src/mem.cpp (hunk 24 above)
Comment (pakmarkthub) [reply src/mem.cpp]: Fixed

File: src/gdsync.cpp (hunk 25 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: Removed the code for now. The busy polling may create a lot of memory traffic for now. Let's improve this later.

File: include/gdsync/core.h (hunk 10 above)
Comment (pakmarkthub) [reply include/gdsync/core.h:185]: Moved peer-related data structures from core.h to mlx5_peer.h

File: src/apis.cpp (hunk 11 above)
Comment (pakmarkthub) [reply src/apis.cpp:93]: Implemented

File: src/apis.cpp (hunk 12 above)
Comment (pakmarkthub) [reply src/apis.cpp]: Moved to mlx5.cpp

File: src/apis.cpp (hunk 13 above)
Comment (pakmarkthub) [reply src/apis.cpp]: Moved to mlx5.cpp

File: src/apis.cpp (hunk 14 above)
Comment (pakmarkthub) [reply src/apis.cpp:390]: Fixed

File: src/apis.cpp (hunk 15 above)
Comment (pakmarkthub) [reply src/apis.cpp:698]: Fixed

File: src/archutils.h (hunk 16 above)
Comment (pakmarkthub) [reply src/archutils.h:1]: Why not?

File: include/gdsync/core.h [hunk 31]
```diff
@@ -113,8 +102,7 @@ struct gds_qp {
 
 /* \brief: Poll a peer-enabled CQ.
  *
- * It works similar to ibv_poll_cq but may return EBUSY if a peek entry
- * associated with the CQE is still busy.
+ * It works similar to ibv_poll_cq while taking peer peek entry into account.
```
Comment (drossetti) [inline include/gdsync/core.h]: preexisting, similar = similarly

Comment (drossetti) [reply include/gdsync/core.h]: "while blocking CPU access to the CQ entries until the GPU had a chance to observe them"

File: src/apis.cpp [hunk 32]
```diff
@@ -747,11 +629,24 @@ int gds_post_descriptors(size_t n_descs, gds_descriptor_t *descs, int flags)
                         case GDS_TAG_SEND: {
                                 gds_dbg("desc[%zu] SEND\n", i);
                                 gds_send_request_t *sreq = desc->send;
+                                retcode = gds_post_ops_on_cpu(sreq->commit.entries, sreq->commit.storage, flags);
+                                if (retcode) {
+                                        gds_err("error %d in gds_post_ops_on_cpu\n", retcode);
+                                        ret = retcode;
+                                        goto out;
+                                }
+                        break;
```
Comment (drossetti) [inline src/apis.cpp]: extra break?

File: src/apis.cpp [hunk 33]
```diff
@@ -747,11 +629,24 @@ int gds_post_descriptors(size_t n_descs, gds_descriptor_t *descs, int flags)
...
                                 break;
                         }
                         case GDS_TAG_WAIT: {
                                 gds_dbg("desc[%zu] WAIT\n", i);
                                 gds_wait_request_t *wreq = desc->wait;
+                                retcode = gds_post_ops_on_cpu(wreq->peek.entries, wreq->peek.storage, flags);
```
Comment (drossetti) [inline src/apis.cpp]: I don't understand, this call to gds_post_ops_on_cpu is already on master code.
how come it is shown here ?

File: include/gdsync/core.h (hunk 27 above)
Comment (drossetti) [reply include/gdsync/core.h:74]: can you wrap the stuff coming from peer-direct async implementation into a mlx5 specific cq struct ?

File: include/gdsync/core.h (hunk 29 above)
Comment (drossetti) [reply include/gdsync/core.h]: move these into a mlx5 specific qp

File: include/gdsync/core.h (hunk 30 above)
Comment (drossetti) [reply include/gdsync/core.h]: this is still here

File: src/mlx5.cpp [hunk 34]
```diff
@@ -330,6 +318,225 @@ int gds_mlx5_get_dword_wait_info(uint32_t *ptr, uint32_t value, int flags, gds_m
...
+
+        struct gds_mlx5_peer_op_wr *wr;
+
+        if (commit->entries < 3) {
+                gds_err("not enough entries in gds_mlx5_peer_commit.\n");
+                ret = EINVAL;
+                goto out;
+        }
+
+        if (p_ewr->opcode != IBV_WR_SEND) {
```
Comment (drossetti) [inline src/mlx5.cpp]: can you file an issue about this?
for comm, we need RDMA_WRITE as well

File: src/mlx5.cpp [hunk 35]
```diff
@@ -330,6 +318,225 @@ int gds_mlx5_get_dword_wait_info(uint32_t *ptr, uint32_t value, int flags, gds_m
...
+        wr->wr.dword_va.data = htonl(qp->sq_cur_post & 0xffff);
+        wr->wr.dword_va.target_id = qp->peer_va_id_dbr;
+        wr->wr.dword_va.offset = sizeof(uint32_t) * MLX5_SND_DBR;
+        wr = wr->next;
+
+        wr->type = GDS_MLX5_PEER_OP_FENCE;
+        wr->wr.fence.fence_flags = GDS_PEER_FENCE_OP_WRITE | GDS_PEER_FENCE_FROM_HCA | GDS_PEER_FENCE_MEM_SYS;
+        wr = wr->next;
+
+        wr->type = GDS_MLX5_PEER_OP_STORE_QWORD;
```
Comment (drossetti) [inline src/mlx5.cpp]: we have extra fixes and optimizations in https://github.com/gpudirect/libmlx5/commits/fixes and https://github.com/gpudirect/libmlx5/commits/copyblock-4.3 which should be pulled here.
would you mind filing a new issue?

File: tests/gds_kernel_latency.c [hunk 36]
```diff
@@ -532,76 +535,69 @@ static int pp_wait_cq(struct pingpong_context *ctx, int is_client)
...
-				 .remote_qkey = 0x11111111
-			 }
-		},
-		.comp_mask = 0
-	};
-#if 0
-	if (IBV_QPT_UD != gds_qpt) {
-		memset(&ewr, 0, sizeof(ewr));
-		ewr.num_sge = 1;
-		ewr.exp_send_flags = IBV_EXP_SEND_SIGNALED;
```
Comment (drossetti) [inline tests/gds_kernel_latency.c:560]: is the SIGNALED flag gone?

File: tests/gds_kernel_latency.c [hunk 37]
```diff
@@ -532,76 +535,69 @@ static int pp_wait_cq(struct pingpong_context *ctx, int is_client)
...
-		.addr	= (uintptr_t) ctx->txbuf,
-		.length = ctx->size,
-		.lkey	= ctx->mr->lkey
-	};
-	gds_send_wr ewr = {
-		.wr_id	    = PINGPONG_SEND_WRID,
-		.sg_list    = &list,
-		.num_sge    = 1,
-		.exp_opcode = IBV_EXP_WR_SEND,
-		.exp_send_flags = IBV_EXP_SEND_SIGNALED,
```
Comment (drossetti) [inline tests/gds_kernel_latency.c:584]: _SIGNALED gone here too

File: tests/gds_kernel_latency.c [hunk 38]
```diff
@@ -532,76 +535,69 @@ static int pp_wait_cq(struct pingpong_context *ctx, int is_client)
...
-				 .remote_qkey = 0x11111111
-			 }
-		},
-		.comp_mask = 0
-	};
-	
-	if (IBV_QPT_UD != gds_qpt) {
-		memset(&ewr, 0, sizeof(ewr));
-		ewr.num_sge = 1;
-		ewr.exp_send_flags = IBV_EXP_SEND_SIGNALED;
```
Comment (drossetti) [inline tests/gds_kernel_latency.c:598]: same here

File: tests/gds_kernel_loopback_latency.c [hunk 39]
```diff
@@ -461,123 +446,113 @@ static int unblock_server_stream(struct pingpong_context *ctx)
...
-		.addr	= (uintptr_t) ctx->txbuf,
-		.length = ctx->size,
-		.lkey	= ctx->mr->lkey
-	};
-	gds_send_wr ewr = {
-		.wr_id	    = PINGPONG_SEND_WRID,
-		.sg_list    = &list,
-		.num_sge    = 1,
-		.exp_opcode = IBV_EXP_WR_SEND,
-		.exp_send_flags = IBV_EXP_SEND_SIGNALED,
```
Comment (drossetti) [inline tests/gds_kernel_loopback_latency.c:515]: same here

**Review by drossetti — CHANGES_REQUESTED** [review]

File: include/gdsync/core.h (hunk 30 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Reverted

File: src/mlx5.cpp (hunk 34 above)
Comment (pakmarkthub) [reply src/mlx5.cpp]: Filed an issue.

File: src/mlx5.cpp (hunk 35 above)
Comment (pakmarkthub) [reply src/mlx5.cpp]: Filed an issue.

File: tests/gds_kernel_latency.c (hunk 36 above)
Comment (pakmarkthub) [reply tests/gds_kernel_latency.c:560]: Added.

File: tests/gds_kernel_latency.c (hunk 37 above)
Comment (pakmarkthub) [reply tests/gds_kernel_latency.c:584]: Added

File: tests/gds_kernel_latency.c (hunk 38 above)
Comment (pakmarkthub) [reply tests/gds_kernel_latency.c:598]: Added

File: tests/gds_kernel_loopback_latency.c (hunk 39 above)
Comment (pakmarkthub) [reply tests/gds_kernel_loopback_latency.c:515]: Added

File: src/apis.cpp (hunk 32 above)
Comment (pakmarkthub) [reply src/apis.cpp]: Removed

File: include/gdsync/core.h (hunk 31 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Fixed

**Comment by pakmarkthub** [comment]
I should have fixed/addressed almost all the comments.

@drossetti Please re-review.

**Comment by pakmarkthub** [comment]
@drossetti 
I made post_send support all types of QP, IBV_WR_\*, and IBV_SEND_\*. Please take a look if you have time.

**Pull request closed without merging**